nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py

```diff
@@ -0,0 +1,478 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from copy import deepcopy
+from typing import Mapping
+
+import numpy as np
+import torch
+import wandb
+from apex.transformer.pipeline_parallel.utils import get_micro_batch_size, get_num_microbatches
+from megatron.core import parallel_state
+from megatron.core.pipeline_parallel.schedules import get_forward_backward_func
+from megatron.core.tensor_parallel.random import get_cuda_rng_tracker, get_data_parallel_rng_tracker_name
+from omegaconf import OmegaConf
+from PIL import Image
+from torch import nn
+from tqdm import tqdm
+
+import nemo.collections.multimodal.parts.stable_diffusion.pipeline as sampling_utils
+from nemo.collections.multimodal.models.text_to_image.stable_diffusion.diffusion_engine import (
+    DiffusionEngine,
+    MegatronDiffusionEngine,
+)
+from nemo.collections.multimodal.models.text_to_image.stable_diffusion.ldm.ddpm import (
+    LatentDiffusion,
+    MegatronLatentDiffusion,
+)
+from nemo.collections.multimodal.parts.stable_diffusion.sdxl_helpers import (
+    do_sample,
+    get_batch,
+    get_unique_embedder_keys_from_conditioner,
+)
+from nemo.collections.multimodal.parts.stable_diffusion.sdxl_pipeline import get_sampler_config
+from nemo.collections.nlp.modules.common.megatron.utils import average_losses_across_data_parallel_group
+from nemo.collections.nlp.parts.utils_funcs import get_last_rank
+from nemo.utils import logging
+from nemo_aligner.models.alignable_interface import SupervisedInterface
+from nemo_aligner.utils.train_utils import (
+    finish_validation_step,
+    grad_reductions,
+    prepare_for_training_step,
+    prepare_for_validation_step,
+)
+from nemo_aligner.utils.utils import (
+    _get_autocast_dtype,
+    adapter_control,
+    configure_batch_sizes,
+    get_iterator_k_split_list,
+)
+
+BatchType = Mapping[str, torch.tensor]
+
+
+def calculate_gaussian_kl_penalty_shared_var(curr_eps, init_eps):
+    diff = curr_eps - init_eps
+    kl = torch.sum(diff ** 2, dim=(1, 2, 3, 4), keepdim=True).flatten()
+    dimensionality = torch.numel(curr_eps[0])
+    kl /= dimensionality
+    return kl
+
+
+class MegatronSDXLDRaFTPModel(MegatronDiffusionEngine, SupervisedInterface):
```

Comment:
@rohitrango Do you mind leaving a comment like
```suggestion

# TODO(@rohitrango): Remove after SD/SDXL APIs merged in NeMo so only one config is needed in Aligner
```
This way we won't forget that to deduplicate this code once Nemo APIs are merged